Rework polygon_ring_appender to avoid slowness

pwicks86 opened the pull request:
<!--
IMPORTANT: Do NOT use GitHub to post any questions or support requests!
           They will be closed immediately and ignored.

Make sure that the title of your commit(s) is descriptive. Typically, they
should be formatted as "component/filename: Describe what the commit does (fixes #ticket)",
so that anyone that parses 'git log' immediately knows what a commit is about.
Do not hesitate to provide more context in the longer part of the commit message.

GOOD: "GTiff: fix wrong color interpretation with -co ALPHA=YES (fixes #1234)

When -co ALPHA=YES was used, but PHOTOMETRIC was not specified, the ExtraSample
tag was wrongly set to unspecified.
"

BAD: "Fix crash", "fix #1234"

In case you need several iterations to make continuous integration happy,
please squash your commits in a single one at the end. See
[Contributing](https://github.com/OSGeo/gdal/blob/master/CONTRIBUTING.md)
-->

## What does this PR do?

This patch reworks `polygon_ring_appender` to avoid some cases that caused
massive slowness. This involved doing 2 main things:

1. The `isIn` check has been reworked to avoid a dependency on OGR and
instead the winding number algorithm has been implemented directly. The
code itself was based on the reference code at geomalgorithms.com

2. Rather than sort all the rings in the destructor, the `addLine`
function has been reworked to put each new ring in the correct position
as it is added. To enable this, the `Ring` class was reworked to hold a
vector of child instances directly, effectively building a tree of
rings. Then in the destructor, we can simply walk the tree and write out
every other level as a polygon.

## What are related issues/pull requests?

#2241

## Tasklist

 - [ ] Add test case(s)
 - [ ] Add documentation
 - [ ] Review
 - [ ] Adjust for comments
 - [ ] All CI builds and checks have passed

## Environment

Provide environment details, if relevant:

* OS: Ubuntu 18.04
* Compiler: gcc 7.5

pwicks86 commented:
Huh, not sure why this fails on MacOS. I don't have an Apple machine to test on at the moment unfortunately.

rouault commented:
> Huh, not sure why this fails on MacOS.

Just look at the logs :-) Linux builds also fail, so this is not platform specific
On ./gdal_unit_test in autotest/cpp (``make gdal_unit_test``):
```
---> group: MarchingSquares:Polygon, test: test<4>
     problem: assertion failed
     failed assertion: "Polygon #1: expected '{ { (2.25,2.5) (2.5,2.25) (3,2.25) (3,2) (3,1.5) (3,1) (3,0.75) (2.5,0.75) (2.25,0.5) (2.25,0) (2,0) (1.5,0) (1,0) (0.75,0) (0.75,0.5) (0.5,0.75) (0,0.75) (0,1) (0,1.5) (0,2) (0,2.25) (0.5,2.25) (0.75,2.5) (0.75,3) (1,3) (1.5,3) (2,3) (2.25,3) (2.25,2.5) } { (1.625,1.5) (1.5,1.625) (1.375,1.5) (1.5,1.375) (1.625,1.5) } } ' actual '{ { (1.625,1.5) (1.5,1.625) (1.375,1.5) (1.5,1.375) (1.625,1.5) } } { { (2.25,2.5) (2.5,2.25) (3,2.25) (3,2) (3,1.5) (3,1) (3,0.75) (2.5,0.75) (2.25,0.5) (2.25,0) (2,0) (1.5,0) (1,0) (0.75,0) (0.75,0.5) (0.5,0.75) (0,0.75) (0,1) (0,1.5) (0,2) (0,2.25) (0.5,2.25) (0.75,2.5) (0.75,3) (1,3) (1.5,3) (2,3) (2.25,3) (2.25,2.5) } } '"

---> group: MarchingSquares:Polygon, test: test<5>
     problem: assertion failed
     failed assertion: "Polygon #1: expected '{ { (4.5,5) (5,5) (5,4.5) (5,4) (5,3.5) (5,3) (5,2.5) (5,2) (5,1.5) (5,1) (5,0.5) (5,0) (4.5,0) (4,0) (3.5,0) (3,0) (2.5,0) (2,0) (1.5,0) (1,0) (0.5,0) (0,0) (0,0.5) (0,1) (0,1.5) (0,2) (0,2.5) (0,3) (0,3.5) (0,4) (0,4.5) (0,5) (0.5,5) (1,5) (1.5,5) (2,5) (2.5,5) (3,5) (3.5,5) (4,5) (4.5,5) } { (4,3.5) (3.5,4) (2.5,4) (1.5,4) (1,3.5) (1,2.5) (1,1.5) (1.5,1) (2.5,1) (3.5,1) (4,1.5) (4,2.5) (4,3.5) } } ' actual '{ { (4,3.5) (3.5,4) (2.5,4) (1.5,4) (1,3.5) (1,2.5) (1,1.5) (1.5,1) (2.5,1) (3.5,1) (4,1.5) (4,2.5) (4,3.5) } } { { (4.5,5) (5,5) (5,4.5) (5,4) (5,3.5) (5,3) (5,2.5) (5,2) (5,1.5) (5,1) (5,0.5) (5,0) (4.5,0) (4,0) (3.5,0) (3,0) (2.5,0) (2,0) (1.5,0) (1,0) (0.5,0) (0,0) (0,0.5) (0,1) (0,1.5) (0,2) (0,2.5) (0,3) (0,3.5) (0,4) (0,4.5) (0,5) (0.5,5) (1,5) (1.5,5) (2,5) (2.5,5) (3,5) (3.5,5) (4,5) (4.5,5) } } '"
```

jratike80 commented:
Out of curiosity, do you have metrics about the speedup?

rouault commented on gdal/alg/marching_squares/point.h:
```diff
@@ -59,6 +59,16 @@ std::ostream & operator<<(std::ostream & o, const Point& p)
+inline int
+isLeft(Point p0, Point p1, Point p2 )
+{
+    return static_cast<int>(( (p1.x - p0.x) * (p2.y - p0.y)
```
casting to int could potentially overflow. Cannot the return be a bool instead and test the floating point value againt > 0 ?

rouault commented on gdal/alg/marching_squares/point.h:
```diff
@@ -59,6 +59,16 @@ std::ostream & operator<<(std::ostream & o, const Point& p)
+// Returns greater than zero for p2 left of the line defined by p0 and p1,
+// 0 if p2 is on the line and less than if p2 is to the right of the line
+inline int
+isLeft(Point p0, Point p1, Point p2 )
```
```suggestion
isLeft(const Point& p0, const Point& p1, const Point& p2 )
```

rouault commented on gdal/alg/marching_squares/polygon_ring_appender.h:
```diff
@@ -51,59 +52,46 @@ class PolygonRingAppender
+            auto p1 = *otherIter;
+            do {
+                otherIter++;
+                auto p2 = *otherIter;
```
I believe there's a risk to dereference otherIter when its already at other.points.end(). The test at the end of this scope (`` } while (...)``) should probably be moved here

rouault commented on gdal/alg/marching_squares/polygon_ring_appender.h:
```diff
@@ -114,8 +102,22 @@ class PolygonRingAppender
 #endif
     };
 
+    void processTree(std::vector<Ring> &tree, int level) {
```
const maybe ?
```suggestion
    void processTree(const std::vector<Ring> &tree, int level) {
```

pwicks86 commented:
> Out of curiosity, do you have metrics about the speedup?

I happened to have some DEM data for the Mt. Shasta area on hand, so I mostly used that for testing while I was working on this since it's a relatively small data set. Before making this change, generating a polygon contour at 10 meter intervals for ned19_n41x50_w122x50_ca_mountshasta_2010.img took about 5 minutes and 30 seconds. With this change, the same contour could be generated in 20 seconds. On a much larger DEM that I needed for another project, a subset of a EU-DEM tile 34439x24561 in size, I was able to generate my needed contour in about 20 minutes compared to several hours at least(I don't have an exact time for the old version, I never let it run to completion)

pwicks86 commented:
> On ./gdal_unit_test in autotest/cpp (make gdal_unit_test):

Ah, I had just been using `pytest` and hadn't seen failures related to this, hadn't realized that there was a separate unit test to be run. I've confirmed I can reproduce that test failure locally, will investigate.

pwicks86 commented on gdal/alg/marching_squares/polygon_ring_appender.h:
```diff
@@ -129,49 +131,69 @@ class PolygonRingAppender
+            // 1. Find all the inner rings of the parent that are inside the new ring
+            auto trueGroupIt = std::partition(
+                parentRing->interiorRings.begin(),
+                parentRing->interiorRings.begin(),
```
Oops, this is a typo, should be `parentRing->interiorRings.end()`, which (I think?) turns this whole section into pretty much a no-op, as the partitioned result will always be of size 0. This might be causing my test failures, investigating.

pwicks86 commented:
Updated the PR in response to the comments above and also fixed the test failure, which I think was a combination of a few things:

- `isLeft` returned bad values sometimes due to the cast. Fixed by removing the cast and just returning a bool instead.
- The logic for inserting rings into the tree wasn't quite right. There was the bug mentioned above and also the fact that I had not checked the case where a ring didn't fit inside any existing rings; in this case, we need to check the existing top level rings and see if any should be inside the new ring.

rouault commented:
With Visual Studio 2015:
```
c:\projects\gdal\gdal\alg\marching_squares\polygon_ring_appender.h(145): error C2039: 'back_inserter': is not a member of 'std' (compiling source file contour.cpp)
```
Probably due to a missing ```#include <iterator>``` ?

rouault commented:
Thanks !

hlthi commented:
> Out of curiosity, do you have metrics about the speedup?

In my case 750 second to 6 second.

Amazing @pwicks86  👏

pwicks86 commented:
@hlthi happy to help :)
